File: components/m5_ain_4_20ma/sensor.py
"""Sensor platform for M5Stack AIN 4-20mA unit.

This component reads 4-20mA current values from the M5Stack Unit AIN
4-20mA (SKU: U162) via I2C.

I2C Address: 0x55
Current Register: 0x20 (2 bytes, little-endian)
"""

import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import i2c, sensor
from esphome.const import CONF_ADDRESS

CONF_SCALING_FACTOR = "scaling_factor"

m5_ain_4_20ma_ns = cg.esphome_ns.namespace("m5_ain_4_20ma")
M5AIN4_20MASensor = m5_ain_4_20ma_ns.class_(
    "M5AIN4_20MASensor", cg.PollingComponent, sensor.Sensor, i2c.I2CDevice
)

CONFIG_SCHEMA = (
    sensor.sensor_schema(
        M5AIN4_20MASensor,
        unit_of_measurement="mA",
        icon="mdi:current-ac",
        accuracy_decimals=2,
        device_class="current",
        state_class="measurement",
    )
    .extend(
        {
            cv.Optional(CONF_SCALING_FACTOR, default=0.01): cv.float_range(
                min=0.0001, max=100.0
            ),
        }
    )
    .extend(cv.polling_component_schema("1s"))
    .extend(i2c.i2c_device_schema(0x55))
)


async def to_code(config):
    var = await sensor.new_sensor(config)
    await cg.register_component(var, config)
    await i2c.register_i2c_device(var, config)

    cg.add(var.set_address(config[CONF_ADDRESS]))
    cg.add(var.set_scaling_factor(config[CONF_SCALING_FACTOR]))

File: components/m5_ain_4_20ma/sensor.cpp
#include "m5_ain_4_20ma.h"
#include "esphome/core/log.h"

namespace esphome {
namespace m5_ain_4_20ma {

static const char *const TAG = "m5_ain_4_20ma";

void M5AIN4_20MASensor::setup() {
  ESP_LOGI(TAG, "Setting up M5Stack AIN 4-20mA sensor");
  ESP_LOGI(TAG, "I2C Address: 0x%02X", this->address_);
  ESP_LOGI(TAG, "Scaling Factor: %.4f", this->scaling_factor_);
}

void M5AIN4_20MASensor::dump_config() {
  LOG_SENSOR("", "M5 AIN 4-20mA", this);
  ESP_LOGCONFIG(TAG, "  I2C Address: 0x%02X", this->address_);
  ESP_LOGCONFIG(TAG, "  Scaling Factor: %.4f", this->scaling_factor_);
}

void M5AIN4_20MASensor::update() {
  // Read current value from register 0x20 (current value register)
  uint8_t data[2];
  
  ESP_LOGV(TAG, "Reading from I2C address 0x%02X", this->address_);
  
  // Write register address
  if (!this->write_bytes(0x20, nullptr, 0)) {
    ESP_LOGW(TAG, "Failed to write register address");
    this->publish_state(NAN);
    return;
  }
  
  // Read 2 bytes (little-endian)
  if (!this->read_bytes(0x20, data, 2)) {
    ESP_LOGW(TAG, "Failed to read data from device at 0x%02X", this->address_);
    this->publish_state(NAN);
    return;
  }
  
  // Combine bytes (little-endian: LSB first)
  uint16_t raw_value = (uint16_t)data[0] | ((uint16_t)data[1] << 8);
  
  ESP_LOGV(TAG, "Raw value: %u (0x%02X 0x%02X)", raw_value, data[0], data[1]);
  
  // Convert to mA using scaling factor
  float current_ma = raw_value * this->scaling_factor_;
  
  ESP_LOGV(TAG, "Current: %.4f mA", current_ma);
  
  // Sanity check for reasonable 4-20mA range
  if (current_ma < 0.0 || current_ma > 30.0) {
    ESP_LOGW(TAG, "Unusual current reading: %.4f mA (raw: %u)", current_ma, raw_value);
    ESP_LOGW(TAG, "You may need to adjust the scaling factor");
    // Still publish the value but log a warning
  }
  
  // Publish the sensor state
  this->publish_state(current_ma);
}

// Setter methods
void M5AIN4_20MASensor::set_address(uint8_t address) {
  this->address_ = address;
}

void M5AIN4_20MASensor::set_scaling_factor(float scaling_factor) {
  this->scaling_factor_ = scaling_factor;
}

}  // namespace m5_ain_4_20ma
}  // namespace esphome